lucene/core/src/java/org/apache/lucene/util/fst/FST.java

```diff
@@ -864,6 +1012,68 @@ private long readUnpackedNodeTarget(BytesReader in) throws IOException {
     return readNextRealArc(arc, in);
   }
 
+  /**
+   * Reads the presence bits of a direct-addressing node, store them in the provided arc {@link Arc#bitTable()} and returns the number of presence bytes.
+   */
+  private int readPresenceBytes(Arc<T> arc, BytesReader in) throws IOException {
+    int numPresenceBytes = getNumPresenceBytes(arc.numArcs());
+    long[] presenceBits = new long[numPresenceBytes + 7 >>> 3];
+    int longIndex = 0;
+    int byteIndex = 0;
+    for (int i = 0; i < numPresenceBytes; i++) {
+      presenceBits[longIndex] |= (long) (in.readByte() & 0xFF) << (i << 3);
+      if (++byteIndex == 8) {
+        byteIndex = 0;
+        longIndex++;
+      }
+    }
+    arc.bitTable = presenceBits;
+    assert checkPresenceBytesAreValid(arc);
+    return numPresenceBytes;
+  }
+
+  static boolean checkPresenceBytesAreValid(Arc arc) {
+    assert (arc.bitTable()[0] & 1L) != 0; // First bit must be set.
+    assert (arc.bitTable()[arc.bitTable().length - 1] & (1L << (arc.numArcs() - 1))) != 0; // Last bit must be set.
+    assert countBits(arc.bitTable()) <= arc.numArcs(); // Total bit set must be <= label range.
+    return true;
+  }
+
+  /**
+   * Counts all bits set in the provided longs.
+   */
+  static int countBits(long[] bits) {
+    int bitCount = 0;
+    for (long l : bits) {
+      bitCount += Long.bitCount(l);
+    }
+    return bitCount;
+  }
+
+  /**
+   * Returns whether the bit at given index is set.
+   */
+  static boolean isBitSet(long[] bits, int bitIndex) {
+    return (bits[bitIndex >> 6] & (1L << (bitIndex & 63))) != 0;
+  }
+
+  /**
+   * Counts the bits set up to the given bit index, exclusive.
+   */
+  static int countBitsUpTo(long[] bits, int bitIndex) {
+    int bitCount = 0;
+    int lastLong = bitIndex >>> 6; // index / 64
+    for (int i = 0; i < lastLong; i++) {
+      bitCount += Long.bitCount(bits[i]);
+    }
+    int indexInLastLong = bitIndex & 63; // index % 64
+    if (indexInLastLong != 0) {
+      long mask = 0xFFFFFFFFFFFFFFFFL >>> (Long.SIZE - indexInLastLong);
+      bitCount += Long.bitCount(bits[lastLong] & mask);
+    }
```

Comment:
we should be able to remove the conditional? e.g. something like this:

```
long mask = (1L << bitIndex) - 1L; // shifts are mode 64
bitCount += Long.bitCount(bits[lastLong] & mask);
```

Reply:
Nice!